src/js/editor/post.js

```diff
@@ -0,0 +1,474 @@
+   *     });
+   *     // Will result some markers possibly being split, and the markup
+   *     // being applied to all markers between the split.
+   *
+   * The return value will be all markers between the split, the same return
+   * value as `splitMarkers`.
+   *
+   * @method applyMarkupToMarkers
+   * @param {Object} markerRange Object with offsets, {headMarker, headOffset, tailMarker, tailOffset}
+   * @param {Object} markup A markup post abstract node
+   * @return {Array} of markers that are inside the split
+   * @public
+   */
+  applyMarkupToMarkers(markerRange, markup) {
+    const markers = this.splitMarkers(markerRange);
+    markers.forEach(marker => {
+      marker.addMarkup(markup);
+      marker.section.renderNode.markDirty();
+    });
+
+    this.scheduleRerender();
+    this.scheduleDidUpdate();
+
+    return markers;
+  }
+
+  /**
+   * Given a markerRange (for example `editor.cursor.offsets`) remove the given
+   * markup from all contained markers. The markup must be provided as a post
+   * abstract node.
+   *
+   * Usage:
+   *
+   *     let markerRange = editor.cursor.offsets;
+   *     let markup = markerRange.headMarker.markups[0];
+   *     editor.run((postEditor) => {
+   *       postEditor.removeMarkupFromMarkers(markerRange, markup);
+   *     });
+   *     // Will result some markers possibly being split, and the markup
+   *     // being removed from all markers between the split.
+   *
+   * The return value will be all markers between the split, the same return
+   * value as `splitMarkers`.
+   *
+   * @method removeMarkupFromMarkers
+   * @param {Object} markerRange Object with offsets, {headMarker, headOffset, tailMarker, tailOffset}
+   * @param {Object} markup A markup post abstract node
+   * @return {Array} of markers that are inside the split
+   * @public
+   */
+  removeMarkupFromMarkers(markerRange, markup) {
+    const markers = this.splitMarkers(markerRange);
+    markers.forEach(marker => {
+      marker.removeMarkup(markup);
+      marker.section.renderNode.markDirty();
+    });
+
+    this.scheduleRerender();
+    this.scheduleDidUpdate();
+
+    return markers;
+  }
+
+  /**
+   * Insert a given section before another one, updating the post abstract
+   * and the rendered UI.
+   *
+   * Usage:
+   *
+   *     let markerRange = editor.cursor.offsets;
+   *     let sectionWithCursor = markerRange.headMarker.section;
+   *     let section = editor.builder.createCardSection('my-image');
+   *     editor.run((postEditor) => {
+   *       postEditor.insertSectionBefore(section, sectionWithCursor);
+   *     });
+   *
+   * @method insertSectionBefore
+   * @param {Object} section The new section
+   * @param {Object} beforeSection The section "before" is relative to
+   * @public
+   */
+  insertSectionBefore(section, beforeSection) {
+    this.editor.post.sections.insertBefore(section, beforeSection);
+    this.editor.post.renderNode.markDirty();
```

Comment:
does the entire post need to be marked dirty? Can we only dirty the section and beforeSection?